rust/datafusion/src/physical_plan/string_expressions.rs

```diff
@@ -34,40 +38,1167 @@ macro_rules! downcast_vec {
+                            } else if x.len() < start_pos {
+                                Ok("")
+                            } else if x.len() < start_pos + count_usize {
+                                Ok(&x[start_pos..])
+                            } else {
+                                Ok(&x[start_pos..start_pos + count_usize])
+                            }
+                        })
+                        .transpose()
+                    }
+                })
+                .collect()
+        }
+        other => Err(DataFusionError::Internal(format!(
+            "substr was called with {} arguments. It requires 2 or 3.",
+            other
+        ))),
+    }
+}
+
+/// Converts the number to its equivalent hexadecimal representation.
+pub fn to_hex_i32(args: &[ArrayRef]) -> Result<GenericStringArray<i32>> {
+    let integer_array: &Int32Array =
+        args[0].as_any().downcast_ref::<Int32Array>().unwrap();
+    // first map is the iterator, second is for the `Option<_>`
+    Ok(integer_array
+        .iter()
+        .map(|x| x.map(|x| format!("{:x}", x)))
+        .collect())
+}
+
+/// Converts the number to its equivalent hexadecimal representation.
+pub fn to_hex_i64(args: &[ArrayRef]) -> Result<GenericStringArray<i64>> {
+    let integer_array: &Int64Array =
+        args[0].as_any().downcast_ref::<Int64Array>().unwrap();
+    // first map is the iterator, second is for the `Option<_>`
+    Ok(integer_array
+        .iter()
+        .map(|x| x.map(|x| format!("{:x}", x)))
+        .collect())
+}
+
+/// Replaces each character in string that matches a character in the from set with the corresponding character in the to set. If from is longer than to, occurrences of the extra characters in from are deleted.
+pub fn translate<T: StringOffsetSizeTrait>(
+    args: &[ArrayRef],
+) -> Result<GenericStringArray<T>> {
+    let string_array: &GenericStringArray<T> = args[0]
+        .as_any()
+        .downcast_ref::<GenericStringArray<T>>()
+        .unwrap();
+
+    let from_array: &GenericStringArray<T> = args[1]
+        .as_any()
+        .downcast_ref::<GenericStringArray<T>>()
+        .unwrap();
+
+    let to_array: &GenericStringArray<T> = args[2]
+        .as_any()
+        .downcast_ref::<GenericStringArray<T>>()
+        .unwrap();
+
+    Ok(string_array
+        .iter()
+        .enumerate()
+        .map(|(i, x)| {
+            if from_array.is_null(i) || to_array.is_null(i) {
+                None
+            } else {
+                x.map(|x: &str| {
+                    let from = from_array.value(i).chars().collect::<Vec<char>>();
+                    // create a hashmap to change from O(n) to O(1) from lookup
+                    let mut from_map: HashMap<char, usize> =
+                        HashMap::with_capacity(from.len());
+                    from.iter().enumerate().for_each(|(index, c)| {
+                        from_map.insert(c.to_owned(), index);
+                    });
```

Comment:
I don't think it really matters but I think you could  build the hashmap by `collecting` the iterator directly. Something like

```suggestion
                    let mut from_map: HashMap<char, usize> =
                      from.iter()
                        .enumerate()
                        .map(|(index, c)| (c.to_owned(), index))
                      .collect();
```

Reply:
this is a good idea as it probably lets the compiler optimize better. Thanks :+1: